lib/jobs/ucs-service-profile-discovery.js

```diff
@@ -0,0 +1,197 @@
+        };
+        this.ucs = new UcsTool();
+        this.ucs.settings = this.settings;
+    }
+    
+    util.inherits(UcsServiceProfileDiscoveryJob, BaseJob);
+
+    /**
+     * @memberOf UcsServiceProfileDiscoveryJob
+     */
+    UcsServiceProfileDiscoveryJob.prototype._run = function() {
+        var self = this;
+        
+        return self.getRoot()
+            .then(function(root) {
+                return [ root, self.createServiceProfile(root) ];
+            })
+            .spread(function() {
+                self._done();
+            })
+            .catch(function(err) {
+                self._done(err);
+            });
+    };
+
+    /**
+     * @function logIn
+     * @description Log into the UCS microservice
+     */
+    UcsServiceProfileDiscoveryJob.prototype.logIn = function () {
+        var username = this.settings.username;
+        var password = this.settings.password;
+        var host = this.settings.ucs;
+        var url= "/login?host=" + host+ "&user="+ username + "&password=" + password;
+        return this.ucs.clientRequest(url)
+            .then(function(response) {
+                return response.body;
+            });
+    };
+
+
+    /**
+     * @function getRoot
+     * @description return all the elements from the Cisco UCS Manager that RackHD cares about
+     */
+    UcsServiceProfileDiscoveryJob.prototype.getRoot = function () {
+        var username = this.settings.username;
+        var password = this.settings.password;
+        var ucs = this.settings.ucs;
+        var url=
+            "/serviceProfile?host=" + ucs+ "&user="+ username + "&password=" + password;
+        return this.ucs.clientRequest(url)
+            .then(function(response) {
+                return response.body;
+            });
+    };
+
+    /**
+     * @function createServiceProfile
+     * @description discovers all the service profiles in the Cisco UCS
+     */
+    UcsServiceProfileDiscoveryJob.prototype.createServiceProfile = function (root) {
+        var self = this;
+        var username = this.settings.username;
+        var password = this.settings.password;
+        var ucs = this.settings.ucs;
+        var baseurl = this.settings.protocol+"://"+this.settings.host+ ":"+this.settings.port;
+        var url=
+            baseurl + "/serviceProfile?host=" + ucs+ "&user="+ username + "&password=" + password;
+        return Promise.try(function () {
+            if (_.has(root, 'ServiceProfile')) {
+                return self.ucs.clientRequest(url)
+                    .then(function(res) {
+                        assert.object(res);
+                        return res.body;
+                    })
+                    .then(function(data) {
+                        assert.object(data);
+                        return data['ServiceProfile']['members']
+                    })
+                    .map(function(members){
+                        self.createNode(members, Constants.NodeTypes.Compute);
```

Comment:
This promise has to be returned, otherwise this function will return earlier before all nodes are really created.